### PR TITLE
Nt sensitive cd

### PR DESCRIPTION
**Pull request by Moroxus**
Non-termination sensitive control dependency algorithm.

### CODE REVIEW

File: tools/llvm-slicer-opts.cpp [hunk 1]
```diff
@@ -127,7 +127,8 @@ SlicerOptions parseSlicerOptions(int argc, char *argv[]) {
         llvm::cl::desc("Choose control dependencies algorithm to use:"),
         llvm::cl::values(
             clEnumValN(dg::CD_ALG::CLASSIC , "classic", "Ferrante's algorithm (default)"),
-            clEnumValN(dg::CD_ALG::CONTROL_EXPRESSION, "ce", "Control expression based (experimental)")
+            clEnumValN(dg::CD_ALG::CONTROL_EXPRESSION, "ce", "Control expression based (experimental)"),
+            clEnumValN(dg::CD_ALG::NTSCD, "ntscd", "Non-termination sensitive control dependencies (implemented for parallel slicing)")
```
Comment (mchalupa) [inline tools/llvm-slicer-opts.cpp]: IMHO it can be used even without the slicing for parallel programs.

File: tools/llvm-blocks-dump.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#include "dg/analysis/PointsTo/PointerAnalysisFI.h"
```
Comment (mchalupa) [inline tools/llvm-blocks-dump.cpp]: A better name would be `llvm-ntscd-dump` or something like that. `llvm-blocks-dump` really does not say what it does.

File: lib/CMakeLists.txt [hunk 3]
```diff
@@ -134,7 +140,8 @@ else()
 	target_link_libraries(LLVMdg
 				PUBLIC LLVMpta
 				PUBLIC LLVMrd
-				PUBLIC ThreadRegions)
+                                PUBLIC ThreadRegions
```
Comment (mchalupa) [inline lib/CMakeLists.txt]: Weird indentation.

File: lib/CMakeLists.txt [hunk 4]
```diff
@@ -125,6 +130,7 @@ if (APPLE)
 				PUBLIC LLVMpta
 				PUBLIC LLVMrd
 				PUBLIC ThreadRegions
+                                PUBLIC ControlDependence
```
Comment (mchalupa) [inline lib/CMakeLists.txt]: Weird indentation.

File: lib/llvm/analysis/ControlDependence/GraphBuilder.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
+#ifndef GRAPHBUILDER_H
+#define GRAPHBUILDER_H
```
[the remark this answers is not in the document]
Comment (Moroxus) [reply lib/llvm/analysis/ControlDependence/GraphBuilder.h]: Yeah, that is good idea. I can fix it. Just one question. Clang shows this warning:

`warning: macro name is a reserved identifier; -Wreserved-id-macro`

All macros starting with underscore and then followed by capital letter are reserved by standard library. Wouldn't it be better to prefix it with `DG_LLVM`? Without leading underscore?

File: lib/llvm/analysis/ControlDependence/GraphBuilder.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+#include <map>
+#include <iosfwd>
```
[the remark this answers is not in the document]
Comment (Moroxus) [reply lib/llvm/analysis/ControlDependence/GraphBuilder.h]: GraphBuilder has methods:

```
void dumpNodes(std::ostream & ostream) const;

void dumpEdges(std::ostream & ostream) const;
```

`<iosfwd>` is just for forward declaration of std::ostream

File: lib/llvm/analysis/ControlDependence/GraphBuilder.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (mchalupa) [reply lib/llvm/analysis/ControlDependence/GraphBuilder.h]: Yes, probably yes.

File: lib/llvm/analysis/ControlDependence/GraphBuilder.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (mchalupa) [reply lib/llvm/analysis/ControlDependence/GraphBuilder.h]: Ok, I over-looked those.

**Comment by mchalupa** [comment]
Merged to master, thanks.

**Pull request closed without merging**